### ISSUE TITLE
lua client.openrom -- how should it work?

### ISSUE THREAD

**Issue by zeromus** [TASEmulators/BizHawk#1141]
opening the rom causes the lua script to reset, which then makes client.openrom run again
what is the intended use of client.openrom? should the script check to see which rom is loaded and then terminate the infinite repeat by skipping the client.openrom? can we enter this in the documentation's new examples system as a demonstration of the intended use?
I see how it's complicated, since a lua script is kind of bound to the life cycle of the emulator core instance, and we don't really want to change that

see initial report #1140 and further discussion #1516

**Comment by GabrielSorensen** [comment]
This is what I have been thinking could be the way in the lua file itself, but this relies on the userbag data stays persistent through the loading of different roms. It's also not pretty, but is manageable.

Just as a suggestion.

<pre>
started = false
if (userdata.containskey("init")) then
	started = userdata.get("Init")
	console.log("userdata retrieved")
else
	userdata.set("Init", false)
	started = userdata.get("Init")
	console.log("Userdata not set")
end
if started then
  Stringstarted = "true";
else
  Stringstarted = "false";
end
console.log("Started?: " .. Stringstarted)
if (started == false) then
	--if lua is loaded, dont rerun lua init
        --do things here that only need to be run once
        <h4><b>client.openrom("rom")</b></h4>
        userdata.set("Init", true) --init has been run once
end
function cleanup()
	userdata.set("Init", false)
end
event.onexit(cleanup())
</pre>
and a more simplified version 
<pre>
started = false
if (userdata.containskey("init")) then
	started = userdata.get("Init")
else
	userdata.set("Init", false)
	started = userdata.get("Init")
end
if (started == false) then
	--if lua is loaded, dont rerun lua init
        --do things here that only need to be run once
        <h4><b>client.openrom("rom")</b></h4>
        userdata.set("Init", true) --init has been run once
end
function cleanup()
	userdata.set("Init", false)
end
event.onexit(cleanup())
</pre>

**Comment by vadosnaprimer** [comment]
Another workaround:

```lua
path = "C:\\Games\\"
game = "SOS (USA)"
file = ".sfc"

if gameinfo.getromname() ~= game then
	print("Opening game: " .. path .. game .. file)
	client.openrom( path .. game .. file )
else
	print("Game opened: "  .. path .. game .. file) 
end

while true do
	emu.frameadvance()
end
```

> To properly fix it there'd need to be a boolean passed all the way to https://github.com/TASVideos/BizHawk/blob/master/BizHawk.Client.EmuHawk/tools/ToolManager.cs#L508 along the lines of `FromLua`.

**Comment by YoshiRulz** [comment]
A nicer version of @GabrielSorensen's script:
```lua
rom_path = "D:\\roms\\SNES\\Super Mario World (U) [!].smc";
if userdata.get("last_openrom_path") == rom_path then
	userdata.remove("last_openrom_path");
else
	userdata.set("last_openrom_path", rom_path);
	client.openrom(rom_path);
end

...
```

**Comment by authorblues** [comment]
Is there any update on this? I can't find any documentation for features being added to allow openrom() to take a parameter, and the issue is still open. It would be very nice to be able to load a rom without the script restarting. It feels like it significantly diminishes the value of being able to load a rom from a script if that causes the script to restart.

**Comment by YoshiRulz** [comment]
Use the above workaround, or migrate your script to an external tool.

**Comment by YoshiRulz** [comment]
Oh, does https://github.com/TASEmulators/BizHawk/commit/dc090b81fe7971fa1838662fc4c56ee35021a2cd address this? I made that change to fix a simple crash: #2940.
